cpp/src/arrow/engine/substrait/function_test.cc

```diff
@@ -0,0 +1,156 @@
+      RecordBatch::Make(schema(std::move(fields)), 1, columns);
+  EXPECT_OK_AND_ASSIGN(std::shared_ptr<Table> table, Table::FromRecordBatches({batch}));
+  return table;
+}
+
+void CheckValidTestCases(const std::vector<ValidFunctionTestCase>& valid_cases) {
+  for (const ValidFunctionTestCase& test_case : valid_cases) {
+    std::shared_ptr<Table> input_table =
+        GetInputTable(test_case.arguments, test_case.data_types);
+    std::cout << "Input Table: " << input_table->ToString() << std::endl;
+    ASSERT_OK_AND_ASSIGN(std::shared_ptr<Buffer> substrait,
+                         internal::CreateScanProjectSubstrait(
+                             test_case.function_id, input_table, test_case.arguments,
+                             test_case.data_types, *test_case.expected_output_type));
+    ASSERT_OK_AND_ASSIGN(auto plan_json, internal::SubstraitToJSON("Plan", *substrait));
+    std::cout << plan_json << std::endl;
+    std::shared_ptr<Table> output_table;
+    std::shared_ptr<compute::SinkNodeConsumer> consumer =
+        std::make_shared<compute::TableSinkNodeConsumer>(&output_table,
+                                                         default_memory_pool());
+
+    // Mock table provider that ignores the table name and returns input_table
+    NamedTableProvider table_provider = [input_table](const std::vector<std::string>&) {
+      std::shared_ptr<compute::ExecNodeOptions> options =
+          std::make_shared<compute::TableSourceNodeOptions>(input_table);
+      return compute::Declaration("table_source", {}, options, "mock_source");
+    };
+
+    ASSERT_OK_AND_ASSIGN(
+        std::shared_ptr<compute::ExecPlan> plan,
+        DeserializePlan(*substrait, std::move(consumer), default_extension_id_registry(),
+                        /*ext_set_out=*/nullptr, std::move(table_provider)));
+    std::cout << "Plan: " << plan->ToString() << std::endl;
+    ASSERT_OK(plan->StartProducing());
+    ASSERT_FINISHES_OK(plan->finished());
+    std::cout << "Output Table: " << output_table->ToString() << std::endl;
+
+    // Could also modify the Substrait plan with an emit to drop the leading columns
+    ASSERT_OK_AND_ASSIGN(output_table,
+                         output_table->SelectColumns({output_table->num_columns() - 1}));
+
+    std::shared_ptr<Table> expected_output =
+        GetOutputTable(test_case.expected_output, test_case.expected_output_type);
+    AssertTablesEqual(*expected_output, *output_table, /*same_chunk_layout=*/false);
+  }
+}
+
+TEST(FunctionMapping, ValidCases) {
+  const std::vector<ValidFunctionTestCase> valid_test_cases = {
+      {{kSubstraitArithmeticFunctionsUri, "add"},
+       {"SILENT", "127", "10"},
+       {nullptr, int8(), int8()},
+       "-119",
+       int8()}};
+  CheckValidTestCases(valid_test_cases);
+  // const std::vector<FunctionTestCase> error_test_cases = {
+  //     {{kSubstraitArithmeticFunctionsUri, "add"},
+  //      {"ERROR", "127", "10"},
+  //      {nullptr, int8(), int8()}}};
```

Comment:
should we remove this?

Reply:
I uncommented this and implemented these cases.